blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java

```diff
@@ -378,49 +388,74 @@ public void parallelStart(@Nonnull FlowNode parallelStartNode, @Nonnull FlowNode
                 pendingActionsForBranches.remove(branchStartNode);
             }
 
-            branch.setPipelineActions(branchActions);
-            // do we have sequential stages for this parallel branch?
+            if (isNodeVisitorDumpEnabled && branchActions.size() > 0) {
+                dump("\t\tAdding " + branchActions.size() + " actions to branch id: " + branch.getId());
+            }
+
+            // Check the stack for this branch, for declarative pipelines it should be non-empty even if only 1 stage
             Stack<FlowNodeWrapper> stack = stackPerEnd.get(endNode.getId());
             if (stack != null && !stack.isEmpty()) {
-                // yes so we can rebuild the graph here
-                // we don't want the first one as it's a duplicate of the parallel parent but with stage type so rid of it
-                FlowNodeWrapper flowNodeWrapper = stack.pop();
+
+                if (isNodeVisitorDumpEnabled) {
+                    dump(String.format("\t\t\"Complex\" stages detected (%d)", stack.size()));
+                }
+
+                FlowNodeWrapper firstNodeWrapper = stack.pop();
+
                 if (stack.isEmpty()) {
+                    // We've got a single non-nested stage for this branch. We're going to discard it from our graph,
+                    // since we use the node for the branch itself, which has been created with the same name...
+
+                    if (isNodeVisitorDumpEnabled) {
+                        dump("\t\tSingle-stage branch");
+                    }
+
+                    // ...but first we need to poach any actions from the stage node we'll be discarding
+                    branchActions.addAll(firstNodeWrapper.getPipelineActions());
```

Comment:
This line is the meat of the fix. The rest is logic cleanup for readability and documentation, and slightly better info when debug mode is on.